src/net/server/mod.rs

```diff
@@ -77,24 +81,110 @@ impl From<axum::http::header::ToStrError> for MpcHelperServerError {
     }
 }
 
+impl From<axum::extract::rejection::PathRejection> for MpcHelperServerError {
+    fn from(err: axum::extract::rejection::PathRejection) -> Self {
+        Self::BadPathString(err.into())
+    }
+}
+
+/// [`From`] implementation for [`MpcServerError::SendError`].
+/// first call `to_string` so as to drop `T` from the [`MpcServerError`]
+impl<T> From<mpsc::error::SendError<T>> for MpcHelperServerError {
+    fn from(err: mpsc::error::SendError<T>) -> Self {
+        Self::SendError(err.to_string().into())
+    }
+}
+
+/// [`From`] implementation for [`MpcServerError::SendError`].
+/// first call `to_string` to as to drop `T` from the [`MpcServerError`]
+impl<T> From<tokio_util::sync::PollSendError<T>> for MpcHelperServerError {
+    fn from(err: tokio_util::sync::PollSendError<T>) -> Self {
+        Self::SendError(err.to_string().into())
+    }
+}
+
 impl IntoResponse for MpcHelperServerError {
     fn into_response(self) -> Response {
         let status_code = match &self {
             Self::BadQueryString(_) | Self::BadPathString(_) | Self::MissingHeader(_) => {
                 StatusCode::UNPROCESSABLE_ENTITY
             }
-            Self::SerdeError(_) | Self::InvalidHeader(_) | Self::WrongBodyLen { .. } => {
-                StatusCode::BAD_REQUEST
-            }
-            Self::HyperError(_) | Self::SendError(_) | Self::BodyAlreadyExtracted(_) => {
-                StatusCode::INTERNAL_SERVER_ERROR
-            }
+            Self::SerdeError(_)
+            | Self::InvalidHeader(_)
+            | Self::WrongBodyLen { .. }
+            | Self::OutOfOrder { .. } => StatusCode::BAD_REQUEST,
+            Self::HyperError(_)
+            | Self::SendError(_)
+            | Self::BodyAlreadyExtracted(_)
+            | Self::MissingExtension(_) => StatusCode::INTERNAL_SERVER_ERROR,
         };
 
         (status_code, self.to_string()).into_response()
     }
 }
 
+/// Provides a mapping of [`QueryId`]s to senders that forward data to a [`Network`]. Every time a
+/// new query is started, a [`Network`] is created to handle communication for that query. When the
+/// server receives a request, it must know which [`Network`] to forward that request to, so it
+/// holds this mapping to accomplish that.
+///
+/// Is shareable by `clone()`ing.
+#[derive(Clone)]
+pub struct MessageSendMap {
+    senders: Arc<Mutex<HashMap<QueryId, mpsc::Sender<MessageChunks>>>>,
+}
+
+impl MessageSendMap {
+    /// returns the sender for a given query
+    /// # Errors
+    /// if sender does not exist
+    /// # Panics
+    /// if lock is already held by current thread
+    pub fn get(
+        &self,
+        query_id: QueryId,
+    ) -> Result<mpsc::Sender<MessageChunks>, MpcHelperServerError> {
+        self.senders.lock().unwrap().get(&query_id).map_or_else(
+            || Err(MpcHelperServerError::query_id_not_found(query_id)),
+            |sender| Ok(sender.clone()),
+        )
+    }
+
+    /// adds a sender for a given query
```

Comment:
I think it should be an error if I attempt to add sender for the same query twice?

Reply:
yeah, that's probably true. implemented that